test/liquidityConversionRate.js

```diff
@@ -869,6 +879,211 @@ contract('kyberReserve for Liquidity', function(accounts) {
+
+            await newLiqConvRatesInst.setLiquidityParams(
+                rInFp,
+                pMinInFp,
+                formulaPrecisionBits,
+                maxCapBuyInWei,
+                maxCapSellInWei,
+                feeInBps,
+                maxSellRateInPrecision,
+                minSellRateInPrecision
+            );
+
+            let reserveInst = await Reserve.new(network, newLiqConvRatesInst.address, admin);
+            await reserveInst.setContracts(network, newLiqConvRatesInst.address, 0);
+
+            await newLiqConvRatesInst.setReserveAddress(reserveInst.address);
+
+            //set reserve balance.
+            let reserveEtherInit = (BigNumber(10).pow(18)).mul(newE0s[id]);
+            await Helper.sendEtherWithPromise(accounts[9], reserveInst.address, reserveEtherInit);
+
+            await reserveInst.approveWithdrawAddress(token.address, accounts[0], true);
+
+            //transfer tokens to reserve.
+            let amount = (BigNumber(10).pow(18)).mul(newT0s[id]);
+            await token.transfer(reserveInst.address, amount.valueOf());
+
+            if (i < newRs.length) {
+                // buy then check sell rate
+                let lastBuyRate = BigNumber(precision).div(newP0s[id]).floor();
+                let srcAmount = maxCapBuyInWei;
+                while (true) {
+                    let bal = await token.balanceOf(reserveInst.address);
+                    if (bal * 1.0 <= precision * 1.0) { break; }
+                    let buyRate = await reserveInst.getConversionRate(ethAddress, token.address, srcAmount, currentBlock);
+                    if (buyRate.valueOf() == 0) {
+                        while (true) {
+                            srcAmount = srcAmount.div(2).floor();
+                            if (srcAmount.valueOf() == 0) { break; }
+                            buyRate = await reserveInst.getConversionRate(ethAddress, token.address, srcAmount, currentBlock);
+                            if (buyRate.valueOf() != 0) { break; }
+                        }
+                        if (buyRate.valueOf() == 0) { break; }
+                    }
+                    lastBuyRate = buyRate;
+
+                    await reserveInst.trade(ethAddress, srcAmount, token.address, user1, buyRate, true, {from:network, value: srcAmount});
+                }
+                let minBuyRate = BigNumber(precision).div(pMax).floor();
+                assert(lastBuyRate * 1.0 >= minBuyRate * 1.0, "rate must be greater than min rate, loop: " + i);
+            } else {
+                // sell then check buy rate
+                let lastSellRate = BigNumber(precision).mul(newP0s[id]).floor();
+                let srcAmount = maxCapSellInWei.div(newP0s[id]).floor();
+                while (true) {
+                    let ethBal = await Helper.getBalancePromise(reserveInst.address);
+                    if (ethBal * 1.0 < precision.div(100) * 1.0) { break; } // less than 0.001 ether
```

Comment:
better not assume precision and token / eth decimals are same

Reply:
Updated using token precision and eth precision